### PR TITLE
Rewrite RedundantVoidReturnRule with SwiftSyntax

### PR DESCRIPTION
**Pull request by marcelofabri**
The correction tests are failing, but wanted to trigger oss-check meanwhile

### CODE REVIEW

**Review by jpsim — APPROVED** [review]
So good. It's hard to write a clean rewriter though. Maybe it's a skill we'll develop over time.

File: Source/SwiftLintFramework/Rules/Idiomatic/RedundantVoidReturnRule.swift [hunk 1]
```diff
@@ -56,49 +65,99 @@ public struct RedundantVoidReturnRule: ConfigurationProviderRule, SubstitutionCo
...
+        let locationConverter: SourceLocationConverter
+        let disabledRegions: [SourceRange]
+
+        init(locationConverter: SourceLocationConverter, disabledRegions: [SourceRange]) {
+            self.locationConverter = locationConverter
+            self.disabledRegions = disabledRegions
         }
 
-        return [match]
+        override func visit(_ node: FunctionSignatureSyntax) -> Syntax {
```
[the remark this answers is not in the document]
Comment (jpsim) [reply Source/SwiftLintFramework/Rules/Idiomatic/RedundantVoidReturnRule.swift:104]: Duplicating the whole thing for `ClosureSignatureSyntax` works but is klunky:

```diff
diff --git a/Source/SwiftLintFramework/Rules/Idiomatic/RedundantVoidReturnRule.swift b/Source/SwiftLintFramework/Rules/Idiomatic/RedundantVoidReturnRule.swift
index 0fd4108252..38d6d7e59a 100644
--- a/Source/SwiftLintFramework/Rules/Idiomatic/RedundantVoidReturnRule.swift
+++ b/Source/SwiftLintFramework/Rules/Idiomatic/RedundantVoidReturnRule.swift
@@ -101,6 +101,25 @@ private extension RedundantVoidReturnRule {
             self.disabledRegions = disabledRegions
         }
 
+        override func visit(_ node: ClosureSignatureSyntax) -> Syntax {
+            guard let output = node.output,
+                  let tokenBeforeOutput = output.previousToken,
+                  output.containsRedundantVoidViolation else {
+                return super.visit(node)
+            }
+
+            let isInDisabledRegion = disabledRegions.contains { region in
+                region.contains(node.positionAfterSkippingLeadingTrivia, locationConverter: locationConverter)
+            }
+
+            guard !isInDisabledRegion else {
+                return super.visit(node)
+            }
+
+            correctionPositions.append(tokenBeforeOutput.endPositionBeforeTrailingTrivia)
+            return super.visit(node.withOutput(nil).removingTrailingSpaceIfNeeded())
+        }
+
         override func visit(_ node: FunctionSignatureSyntax) -> Syntax {
             guard let output = node.output,
                   let tokenBeforeOutput = output.previousToken,
@@ -136,6 +155,32 @@ private extension ReturnClauseSyntax {
     }
 }
 
+private extension ClosureSignatureSyntax {
+    /// `withOutput(nil)` adds a `.spaces(1)` trailing trivia, but we don't always want it.
+    func removingTrailingSpaceIfNeeded() -> Self {
+        guard let trivia = trailingTrivia, let nextToken = nextToken else {
+            return self
+        }
+
+        let containsNewLine = nextToken.leadingTrivia.contains { piece in
+            switch piece {
+            case .newlines:
+                return true
+            default:
+                return false
+            }
+        }
+
+        guard containsNewLine else {
+            return self
+        }
+
+        return withTrailingTrivia(
+            Trivia(pieces: Array(trivia.dropFirst()))
+        )
+    }
+}
+
 private extension FunctionSignatureSyntax {
     /// `withOutput(nil)` adds a `.spaces(1)` trailing trivia, but we don't always want it.
     func removingTrailingSpaceIfNeeded() -> Self {
```

**Comment by jpsim** [comment]
I found a false positive in Lyft's codebase with this:

```swift
let foo = { () -> Void in
  bar.baz(to: blorp)
}
```

The type inference fails here. I think this is an issue with `@discardableResult`. Maybe it's only safe to report violations with functions?

**Comment by marcelofabri** [comment]
> I found a false positive in Lyft's codebase with this:
> 
> ```swift
> let foo = { () -> Void in
>   bar.baz(to: blorp)
> }
> ```
> 
> The type inference fails here. I think this is an issue with `@discardableResult`. Maybe it's only safe to report violations with functions?

isn't this something that the existing `void_return` triggers?

**Comment by jpsim** [comment]
I hope you don't mind @marcelofabri I rebased and updated this branch, it'd be good to get this in.

**Comment by jpsim** [comment]
> Linting Wire with this PR took 1.12s vs 3.55s on main (68% faster)

🤩

**Comment by jpsim** [comment]
I’m considering making the extended coverage opt-out since when used in closures, this isn’t strictly redundant since it helps the type checker.

**Comment by jpsim** [comment]
When running OSSCheck with `include_closures: false`, there are only two differences with the previous version of the rule and they're both false negatives that were previously missed:

- [ ] This PR introduced a violation in Swift: [/validation-test/IDE/crashers_2_fixed/rdar64265821.swift:4:47: warning: Redundant Void Return Violation: Returning Void in a function declaration is redundant (redundant_void_return)](https://github.com/apple/swift/blob/f0082a8a2ffa26264dc80b52cff24bbe59e5adb5/validation-test/IDE/crashers_2_fixed/rdar64265821.swift#L4:47)
- [ ] This PR introduced a violation in Swift: [/validation-test/IDE/crashers_2_fixed/rdar78017503.swift:5:27: warning: Redundant Void Return Violation: Returning Void in a function declaration is redundant (redundant_void_return)](https://github.com/apple/swift/blob/f0082a8a2ffa26264dc80b52cff24bbe59e5adb5/validation-test/IDE/crashers_2_fixed/rdar78017503.swift#L5:27)

**Pull request merged**